pkg/controllers/workspace_controller.go

```diff
@@ -198,22 +227,41 @@ func (c *WorkspaceReconciler) applyWorkspaceResource(ctx context.Context, wObj *
 		return err
 	}
 
-	// Find all nodes that match the labelSelector and instanceType, they are not necessarily created by machines.
+	if featuregates.FeatureGates[consts.FeatureFlagKarpenter] {
+		// Check nodeClaims only if the k8s version is greater than 1.29. As nodeClaim is supported from 1.29.
+		if c.KubernetesVersion.Major >= consts.KubernetesSupportedMajorVersion {
+			// Wait for pending nodeClaims if any before we decide whether to create new node or not.
+			if err := nodeclaim.WaitForPendingNodeClaims(ctx, wObj, c.Client); err != nil {
+				return err
+			}
+		}
+	}
+
+	// Find all nodes that match the labelSelector and instanceType, they are not necessarily created by machines/nodeClaims.
 	validNodes, err := c.getAllQualifiedNodes(ctx, wObj)
 	if err != nil {
 		return err
 	}
 
-	selectedNodes := selectWorkspaceNodes(validNodes, wObj.Resource.PreferredNodes, wObj.Status.WorkerNodes, lo.FromPtr(wObj.Resource.Count))
+	selectedNodes := c.selectWorkspaceNodes(validNodes, wObj.Resource.PreferredNodes, wObj.Status.WorkerNodes, lo.FromPtr(wObj.Resource.Count))
 
 	newNodesCount := lo.FromPtr(wObj.Resource.Count) - len(selectedNodes)
 
 	if newNodesCount > 0 {
 		klog.InfoS("need to create more nodes", "NodeCount", newNodesCount)
-		if err := c.updateStatusConditionIfNotMatch(ctx, wObj, kaitov1alpha1.WorkspaceConditionTypeMachineStatus, metav1.ConditionUnknown,
-			"CreateMachinePending", fmt.Sprintf("creating %d machines", newNodesCount)); err != nil {
-			klog.ErrorS(err, "failed to update workspace status", "workspace", klog.KObj(wObj))
-			return err
+		if featuregates.FeatureGates[consts.FeatureFlagKarpenter] &&
```

Comment:
The logic seems to be wrong. If featuregate is enabled, we should go to the else path, right?

Reply:
The feature flag condition should go to the else. Otherwise, the logic is correct.